core/trashcan.js

```diff
@@ -335,10 +402,80 @@ Blockly.Trashcan.prototype.close = function() {
  * Inspect the contents of the trash.
  */
 Blockly.Trashcan.prototype.click = function() {
-  var dx = this.workspace_.startScrollX - this.workspace_.scrollX;
-  var dy = this.workspace_.startScrollY - this.workspace_.scrollY;
-  if (Math.sqrt(dx * dx + dy * dy) > Blockly.DRAG_RADIUS) {
+  if (!this.hasBlocks) {
     return;
   }
-  console.log('TODO: Inspect trash.');
+
+  var xml = [];
+  for (var i = 0, text; text = this.contents[i]; i++) {
+    xml[i] = Blockly.Xml.textToDom(text).firstChild;
+  }
+  this.flyout_.show(xml);
+};
+
+/**
+ * Indicate that the trashcan can be clicked (by opening it) if it has blocks.
+ * @private
+ */
+Blockly.Trashcan.prototype.mouseOver_ = function() {
+  if (!this.hasBlocks) {
+    return;
+  }
+  this.setOpen_(true);
+};
+
+/**
+ * Close the lid of the trashcan if it was open (Vis. it was indicating it had
+ *    blocks).
+ * @private
+ */
+Blockly.Trashcan.prototype.mouseOut_ = function() {
+  // No need to do a .hasBlocks check here because if it doesn't the trashcan
+  // wont be open in the first place, and setOpen_ won't run.
+  this.setOpen_(false);
+};
+
+/**
+ * Handle a BLOCK_DELETE event. Adds deleted blocks oldXml to the content array.
+ * @returns {!Function} Function to call when a block is deleted.
+ * @private
+ */
+Blockly.Trashcan.prototype.onDelete_ = function() {
+  var trashcan = this;
+  return function(event) {
+    if (!trashcan.MAX_CONTENTS) {
+      return;
+    }
+    if (event.type == Blockly.Events.BLOCK_DELETE) {
+      var textBlock = trashcan.blockXmlToText(event.oldXml);
+      if (trashcan.contents.indexOf(textBlock) != -1) {
+        return;
+      }
+      trashcan.contents.unshift(textBlock);
+      if (trashcan.contents.length > trashcan.MAX_CONTENTS) {
+        trashcan.contents.splice(trashcan.MAX_CONTENTS,
+            trashcan.contents.length - trashcan.MAX_CONTENTS);
+      }
+
+      trashcan.hasBlocks = true;
+      trashcan.minOpenness_ = trashcan.HAS_BLOCKS_LID_ANGLE;
+      trashcan.setLidAngle_(trashcan.minOpenness_ * 45);
+    }
+  };
+};
+
+/**
+ * Converts xml representing a block into text that can be stored in the
+ *    content array.
+ * @param {!Element} xml An XML tree defining the block and any
+ *    connected child blocks.
+ * @returns {string} The XML tree converted into a usable text string.
+ */
+Blockly.Trashcan.prototype.blockXmlToText = function(xml) {
```

Comment:
A couple of comments here:
- You're removing the ID on the root block, but not on any descendants.  That means that you can get multiply copies of the same block group in the flyout, like this:
![image](https://user-images.githubusercontent.com/13686399/49833091-5b3f8a00-fd4d-11e8-8b8c-bfb89fd3fc4c.png)

- You're deleting the next block(s).  We'd like to keep it.  Deleting a block stack and wanting to get the entire stack back seems like a common use case.